docs/2018-11-08-Configuration-Splitting-Simplification.md

```diff
@@ -55,101 +85,106 @@ We want to rewrite `φ(X) ∧ (¬ ∃ Z. α(Z))` part of the pattern above
 to something more manageable, preferably something that does not use `not`
 and `exists`, except in cases where it can be handled by an SMT solver.
 
-We will assume it is provable that
-1. `φ(X) = t(X) ∧ p(X)` where `p(X)` is a predicate,
-   and `t(X)` is a function-like term. It may or may not unify with `α(Z)`.
-1. `α(Z) =  s(Z) ∧ q(Z)` where `s(Z)` is a functional term,
-   composed out of constructor-like symbols and variables
-   and `q(x)` is a predicate (so `α(Z)` is function-like).
+We assume that `φ(X)` and `α(Z)` are function-like formulae.
 
-The proposed solution should not return a broken result when these assumptions
-do not hold, but it is not forced to fully simplify the given term.
+### Justification
 
-Algorithm
----------
-
-Input:
+```
+φ(X) ∧ (¬ ∃ Z. α(Z))
+    =  φ(X) ∧ (¬ ∃ Z. α(Z))
+    =  φ(X) ∧ ⌈φ(X) ∧ (¬ ∃ Z. α(Z))⌉ -- since φ(x) is function-like
+    =  φ(X) ∧ ⌈φ(X) ∧ (∀ Z. ¬α(Z))⌉  -- ¬ ∃ = ∀ ¬
+    =  φ(X) ∧ ⌈∀ Z. (φ(X) ∧ ¬α(Z))⌉  -- since φt(x) does not depend on Z
+    =  φ(X) ∧ (∀ Z. ⌈φ(X) ∧ ¬α(Z)⌉)  -- proven below
+    =  ∀ Z. φ(X) ∧ ⌈φ(X) ∧ ¬α(Z)⌉  -- FOL
+    =  ∀ Z. (φ(X) ∧ ⌈φ(X)⌉ ∧ ¬⌈φ(X) ∧ α(Z)⌉) -- proven below
+    =  ∀ Z. (φ(X) ∧ ¬⌈φ(X) ∧ α(Z)⌉)
+    =  φ(X) ∧ (∀ Z. ¬⌈φ(X) ∧ α(Z)⌉)
+    =  φ(X) ∧ (¬ ∃ Z. ⌈φ(X) ∧ α(Z)⌉)
+```
 
-1. `t(X)` and `p(X)` such that
 
-   - `φ(X)` is `t(X) ∧ p(X)`
-   - `t(X)` is a function-like term
-   - `p(X)` is a predicate
+### Missing details
 
-1. `Z`, `s(Z)` and `q(Z)` such that
 
-   - `α(Z) =  s(Z) ∧ q(Z)`
-   - `s(Z)` is a functional term made of constructor-like symbols and variables
-   - `q(Z)` is a predicate
+#### `⌈φ ∧  ∀ Z.α(Z)⌉ = ∀ Z. ⌈φ ∧ α(Z)⌉`
 
+for any function-like formula `φ` and for any ML formula `α(Z)`.
 
-Output: `t'(X)` and `p'(X)` such that
+__Proof:__
+```
+⌈φ ∧  ∀ Z.α(Z)⌉
+    = ∃ x. x ∈  (φ ∧  ∀ Z.α(Z))
+    = ∃ x. x ∈  φ ∧ x ∈  (∀ Z.α(Z)) // distributivity of ∧ 
+    = ∃ x. x = φ ∧ x ∈  (∀ Z.α(Z))  // φ is function-like
+    = ∃ x. x = φ ∧ ∀ Z.x ∈  α(Z)  // properties of membership
+    = ∃ x. x = φ ∧ ∀ Z.φ ∈  α(Z) // substitution
+    = ⌈φ⌉ ∧ ∀ Z.φ ∈  α(Z) // properties of ⌈_⌉
+    = ∀ Z.⌈φ⌉ ∧ φ ∈  α(Z) 
+    = ∀ Z.(∃ x. x = φ) ∧ φ ∈  α(Z)  // properties of ⌈_⌉
```

Comment:
If you're expanding the `ceil` in `(∃ x. x = φ)`, why bother converting the exists-equality in a `ceil` two lines above?